rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -818,1663 +791,2181 @@ private Type inferPathExprType(PathExpr pe, TypePath path) {
+ * complex root type allowed for `self` parameters, such as `&`, `Box`, `Rc`,
+ * `Arc`, and `Pin`.
+ *
+ * We strip away the complex root type for performance reasons only, which will
+ * allow us to construct a much smaller set of candidate call targets (otherwise,
+ * for example _a lot_ of methods have a `self` parameter with a `&` root type).
+ */
+bindingset[path, type]
+private predicate isComplexRootStripped(TypePath path, Type type) {
+  path.isEmpty() and
+  not complexSelfRoot(type, _)
+  or
+  exists(TypeParameter tp |
+    complexSelfRoot(_, tp) and
+    path = TypePath::singleton(tp) and
+    exists(type)
+  )
+}
+
+/**
+ * Provides logic for resolving calls to methods.
+ *
+ * When resolving a method call, a list of [candidate receiver types][1] is constructed
+ *
+ * > by repeatedly dereferencing the receiver expression's type, adding each type
+ * > encountered to the list, then finally attempting an unsized coercion at the end,
+ * > and adding the result type if that is successful.
+ * >
+ * > Then, for each candidate `T`, add `&T` and `&mut T` to the list immediately after `T`.
+ *
+ * We do not currently model unsized coercions, and we do not yet model the `Deref` trait,
+ * instead we limit dereferencing to standard dereferencing and the fact that `String`
+ * dereferences to `str`.
+ *
+ * Instead of constructing the full list of candidate receiver types
+ *
+ * ```
+ * T1, &T1, &mut T1, ..., Tn, &Tn, &mut Tn
+ * ```
+ *
+ * we recursively compute a set of candidates, only adding a new candidate receiver type
+ * to the set when we can rule out that the method cannot be found for the current
+ * candidate:
+ *
+ * ```text
+ * forall method:
+ *   not current_candidate matches method
+ * ```
+ *
+ * Care must be taken to ensure that the `not current_candidate matches method` check is
+ * monotonic, which we achieve using the monotonic `isNotInstantiationOf` predicate.
+ *
+ * [1]: https://doc.rust-lang.org/reference/expressions/method-call-expr.html#r-expr.method.candidate-receivers
+ */
+private module MethodResolution {
+  /**
+   * Holds if method `m` with the name `name` and the arity `arity` exists in
+   * `i`, and the type of the `self` parameter is `selfType`.
+   *
+   * `strippedTypePath` points to the type `strippedType` inside `selfType`,
+   * which is the (possibly complex-stripped) root type of `selfType`. For example,
+   * if `m` has a `&self` parameter, then `strippedTypePath` is `TRefTypeParameter()`
+   * and `strippedType` is the type inside the reference.
+   */
+  pragma[nomagic]
+  private predicate methodInfo(
+    Method m, string name, int arity, ImplOrTraitItemNode i, AssocFunctionType selfType,
+    TypePath strippedTypePath, Type strippedType
+  ) {
+    exists(FunctionTypePosition pos |
+      assocFunctionInfo(m, name, arity, i, pos, selfType) and
+      strippedType = selfType.getTypeAt(strippedTypePath) and
+      isComplexRootStripped(strippedTypePath, strippedType) and
```

Comment:
Does this work if we are in fact adding a method to one of the types that can appear for self, like `Box` or `Pin`?